course5/homework10/llist.h

```diff
@@ -0,0 +1,184 @@
+            iterator& operator++() 
+            {
+                if (node_ptr != nullptr)
+                {
+                    node_ptr = node_ptr->next.get();
+                    return *this;
+                }
+                else
+                {
+                    throw std::range_error("End of list");
+                }
+            }
+            T& operator*() { return node_ptr->value; }
+            node* node_ptr;
+        };*/
+        template<class P, class Q>
+        struct iterator_class
+        {
+            iterator_class() : node_ptr(nullptr) { }
+            iterator_class(node* ptr) : node_ptr(ptr) {}
+            iterator_class(const iterator_class& other) : node_ptr(other.node_ptr) { }
+            iterator_class(iterator_class&& other) : node_ptr(other.node_ptr) { }
+            iterator_class& operator=(const iterator_class& other) { return other.node_ptr; }
+            iterator_class& operator=(iterator_class&& other) { return other.node_ptr; }
+            bool operator==(const iterator_class& other) { return node_ptr == other.node_ptr; }
+            bool operator!=(const iterator_class& other) { return node_ptr != other.node_ptr; }
+            iterator_class& operator++() 
+            {
+                if (node_ptr != nullptr)
+                {
+                    node_ptr = node_ptr->next.get();
+                    return *this;
+                }
+                else
+                {
+                    throw std::range_error("End of list");
+                }
+            }
+            Q& operator*() { return node_ptr->value; }
+            P node_ptr;
+        };
+        using iterator = iterator_class<node*, T>;
+        using const_iterator = iterator_class<const node*, const T>;
+
+        llist() : len(0) {}
+        llist(llist&& other) : len( other.size() ) { first = std::move( other.first ); }
+        llist& operator=( llist&& other )
+        {
+            len = other.size();
+            first = std::move( other.first );
+            return *this;
+        }
+
+
+        llist(const llist& other) : len( other.size() )
+        {
+            first = std::make_unique<node>();
+            node* new_node = first.get();
+
+            for( auto i = other.begin(); i != other.end(); ++i )
+            {
+                new_node->value = *i;
+                new_node->next = std::make_unique<node>();
+                new_node = new_node->next.get();
+            }
+        }
+
+        llist& operator=( const llist& other )
+        {
+            len = other.size();
+            first = std::make_unique<node>();
+            node* new_node = first.get();
+
+            for( auto i = other.begin(); i != other.end(); ++i )
+            {
+                new_node->value = *i;
+                new_node->next = std::make_unique<node>();
+                new_node = new_node->next.get();
+            }
+            return *this;
+        }
+
+        template<typename Q>
```

Comment:
In case like this, it's better to use some descriptive name for the type, rather than Q. Maybe ```Iterator``` or '''It``` ?